batch/batch/worker/worker.py

```diff
@@ -839,9 +839,19 @@ async def delete_container(self):
         if self.container_is_running():
             try:
                 log.info(f'{self} container is still running, killing crun process')
-                self.process.terminate()
-                self.process = None
-                await check_exec_output('crun', 'kill', '--all', self.container_name, 'SIGTERM')
+
+                try:
+                    await check_exec_output('crun', 'state', self.container_name)
+                except CalledProcessError:
+                    log.info(f'container {self} cannot be killed because it does not exist', exc_info=True)
```

Comment:
Wouldn't we want this branch to get `terminate`d too? Is there a possibility that the crun process is still setting up the container, so it doesn't exist yet, then we set `self.process = None` which leaves a zombie process?

With that in mind, it pretty much makes the except blocks here identical, and we `log.info` regardless in both cases. `crun state` erroring out is essentially telling us the same information as `crun kill` erroring out, it just perhaps feels kinder. I also think these nested try blocks are starting to get confusing.

I see this code as essentially the same, albeit with slightly less clean error messages:

```python
try:
    await check_exec_output('crun', 'state', self.container_name)
    await check_exec_output('crun', 'kill', '--all', self.container_name, 'SIGKILL')
except CalledProcessError:
    log.info(f'Could not delete container {self} because it does not exist', exc_info=True)
    self.process.terminate()
finally:
    self.process = None
```

The exc_info here could still tell us which one failed, but calling state starts to look not quite as useful as before. For the most part what we've done is change the order of `self.process.terminate`, which I think is much better now, and downgraded the error from before to info.

Reply:
I'm still unsure of what I think the best approach is here, but I do think that as long as we're positive there are no leaks, we'll hit delete-nonexistent-container errors that we will have to ignore. I'm alright with trying out something like the above, though I'm unsure how to verify that the state of the system has improved and we're not just silencing errors.

We could have a monitoring loop that ensures the output of `crun list` approximately matches in length the number of containers we expect are running.

Reply:
We might also use something like an `asyncio.wait_for` with a reasonable timeout and wait for `process` to exit. If it doesn't do so within a few seconds then we log an `exception`. This might be the easiest approach to make sure that we see something if it's really borked.